File: src/main/java/io/github/archipelagominecraft/core/api/ArchipelagoLocationType.java
package io.github.archipelagominecraft.core.api;

import com.mojang.serialization.Codec;
import net.minecraft.resources.ResourceLocation;

public interface ArchipelagoLocationType<S extends ArchipelagoLocationType<S, T>,T> {
    ResourceLocation getId();

    void registerLocation(ArchipelagoLocationView<S> archipelagoID, T itemData);

    Codec<T> getDataCodec();
}

File: src/main/kotlin/io/github/archipelagominecraft/core/api/ArchipelagoLocationType.kt
package io.github.archipelagominecraft.core.api

import com.mojang.serialization.Codec


interface ArchipelagoLocationType<S : ArchipelagoLocationType<S, T>, T> {
    val namespace: String
    val typeId: String
    val id: String
        get() = this.namespace + ":" + this.typeId

    fun registerLocation(locationView: ArchipelagoLocationView<S>, itemData: T)

    val dataCodec: Codec<T>
}
